demo-kotlin/app/src/gms/java/com/example/placesdemo/CurrentPlaceTestActivity.kt

```diff
@@ -60,89 +70,263 @@ class CurrentPlaceTestActivity : AppCompatActivity() {
+            )
             != PackageManager.PERMISSION_GRANTED
-            || ContextCompat.checkSelfPermission(this, permission.ACCESS_FINE_LOCATION)
-            != PackageManager.PERMISSION_GRANTED) {
+        ) {
             Toast.makeText(
                 this,
-                "Both ACCESS_WIFI_STATE & ACCESS_FINE_LOCATION permissions are required",
-                Toast.LENGTH_SHORT)
-                .show()
+                "Either ACCESS_FINE_LOCATION or ACCESS_COARSE_LOCATION permission is required.",
+                Toast.LENGTH_SHORT
+            ).show()
+        }
+
+        // 1. Check if either ACCESS_FINE_LOCATION or ACCESS_COARSE_LOCATION is granted. If so,
+        // proceed with finding the current place.
+        if (hasOnePermissionGranted(
+                ACCESS_FINE_LOCATION,
+                ACCESS_COARSE_LOCATION
+            )
+        ) {
+            Log.d(TAG, "Location permission granted. Getting current place.")
+            findCurrentPlaceWithPermissions()
+            return
         }
 
-        // Note that it is not possible to request a normal (non-dangerous) permission from
-        // ActivityCompat.requestPermissions(), which is why the checkPermission() only checks if
-        // ACCESS_FINE_LOCATION is granted. It is still possible to check whether a normal permission
-        // is granted or not using ContextCompat.checkSelfPermission().
-        if (ContextCompat.checkSelfPermission(this, permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
-            ActivityCompat.requestPermissions(this, arrayOf(permission.ACCESS_FINE_LOCATION), 0)
+        // 2. If either permission is not granted, check if a permission rationale dialog must be
+        // shown
+        if (shouldShowPermissionRationale(
+                ACCESS_FINE_LOCATION,
+                ACCESS_COARSE_LOCATION
+            )
+        ) {
+            Log.d(TAG, "Showing permission rationale dialog")
+            RationaleDialog.newInstance(
+                PERMISSION_REQUEST_CODE, true
+            ).show(supportFragmentManager, "dialog")
             return
         }
-        findCurrentPlaceWithPermissions()
+
+        // 3. Otherwise, request permission
+        Log.d(
+            TAG,
+            "No location permission granted. Request permission from the user."
+        )
+        ActivityCompat
+            .requestPermissions(
+                this,
+                arrayOf(
+                    ACCESS_FINE_LOCATION,
+                    ACCESS_COARSE_LOCATION
+                ),
+                PERMISSION_REQUEST_CODE
+            )
     }
 
     /**
      * Fetches a list of [PlaceLikelihood] instances that represent the Places the user is
      * most likely to be at currently.
      */
-    @RequiresPermission(allOf = [permission.ACCESS_FINE_LOCATION, permission.ACCESS_WIFI_STATE])
+    @RequiresPermission(anyOf = [ACCESS_FINE_LOCATION, ACCESS_FINE_LOCATION])
```

Comment:
First one should be ACCESS_COARSE_LOCATION

Reply:
Good catch. Addressed in c945005